### ISSUE TITLE
Gracefully handling errors when the server is down

### ISSUE THREAD

**Issue by ibasin** [paulcbetts/ModernHttpClient#127]
I am using ModernHttpClient for my Xamarin project and it is working great for me. However, I ran into something that seems to be either a bug or an issue I don't understand. If you could address this, it would be much appreciated.

When I make the following call  var data = await httpClient.GetAsync(url); but the server at the url is not running, I get a "System.Net.WebException: The operation couldn’t be completed. (NSURLErrorDomain error -1004.)." However, the exception is thrown from somewhere where I can't catch it. If I try to wrap this call in try catch, the exception is not caught. It appears to be thrown from somewhere else and I have no way of catching and handling it gracefully.

You can try the following code. Put a breakpoint on the throw statement. The code never reaches the trow statement even though System.Net.WebException is thrown from somewhere (I wonder from where). 

            try

            {

                //Call a non-existing/non-working web server 

                var data = await httpClient.GetAsync("http://www.ashdgfajhsdfgkajsf.com");

            }

            catch (Exception)

            {

                throw;

            }

Please help. I need a way to gracefully handle a situation when the server is not running or is inaccessible.

Please let me know if you have any questions or need any additional information.

Thank you very much in advance.

**Comment by paulcbetts** [comment]
> However, the exception is thrown from somewhere where I can't catch it.

Call stack?

**Comment by ibasin** [comment]
I am getting [External Code] in Call Stack.

In the source debugger I am getting: 

Source Not Available
Frame not in module
The current stack frame was not found in a loaded module. Source cannot be shown for this location.

Please see the attached screenshot:

![screen shot 2015-03-12 at 2 31 29 pm](https://cloud.githubusercontent.com/assets/4512592/6625207/e5a15b86-c8c4-11e4-9506-91b87e423e0b.png)

Here is the complete code I am running to reproduce the issue:

            try
            {
                var httpClient = new HttpClient(new NativeMessageHandler());
                //Call a non-existing/non-working web server 
                var data = await httpClient.GetAsync("http://www.ashdgfajhsdfgkajsf.com");
            }
            catch (Exception)
            {
                throw;
            }

**Comment by Yortw** [comment]
What happens if you click continue at this point, does it hit the breakpoint then? Do you have break on all exceptions turned on?

**Comment by ibasin** [comment]
You are right. If I click continue, it hits my catch statement after breaking for a few more exceptions being thrown (I did have break on all exceptions turned on). I can't believe I did not see something so simple. I feel stupid for even asking. Thank you very much! Sorry to bother you.

**Comment by paulcbetts** [comment]
Disable "Just My Code" in debugger options or right click the "External Code" frame

**Comment by paulcbetts** [comment]
@ibasin No worries! I'm just happy when people are successful with ModernHttpClient, you don't need to have Irrefutable Proof™ that it's a bug, questions are welcome here.

**Comment by Yortw** [comment]
No problem. TBH I suggested that because I've seen this a lot in GitHub issues recently. I'm not sure why, perhaps it's just easy to forget or perhaps something has changed in the way VS handles breaking on exceptions. I'm sure I used to turn on break on all exception but not see ones in third party code, and I don't recall changing anything. Disabling 'just my code' as Paul suggested might help, but from the name of that option it sounds like it should be 'enabled' (i.e break on exceptions in just my code, not 1st/3rd party libraries).

Anyway, glad it's solved and Paul isn't chasing a red herring.
